src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2581,6 +2581,40 @@ module Crystal
       false
     end
 
+    def visit(node : OffsetOf)
+      @in_type_args += 1
+      node.offsetof_type.accept self
+      @in_type_args -= 1
+
+      type = node.offsetof_type.type?
+
+      node.offsetof_type.raise "type #{type} can't have instance variables" unless type.is_a?(InstanceVarContainer)
+      node.offsetof_type.raise "can't use typeof inside offsetof expression" if node.offsetof_type.is_a?(TypeOf)
+
+      instance_vars = type.all_instance_vars
+      ivar_name = node.instance_var.as(InstanceVar).name
+      ivar_index = instance_vars.keys.index(ivar_name)
+
+      node.instance_var.raise "type #{type} doesn't have an instance variable called #{ivar_name}" unless ivar_index
+      node.offsetof_type.raise "can't take offsetof element #{ivar_name} of uninstantiated generic type #{type}" if type.is_a?(GenericType)
+
+      if type && type.struct?
+        expanded = NumberLiteral.new(@program.offset_of(type.sizeof_type, ivar_index).to_s, :i32)
+        expanded.type = @program.int32
+        node.expanded = expanded
+      elsif type && type.instance_type.devirtualize.class?
+        expanded = NumberLiteral.new(@program.instance_offset_of(type.sizeof_type, ivar_index).to_s, :i32)
+        expanded.type = @program.int32
+        node.expanded = expanded
```

Comment:
The only difference between these two branches is:

```crystal
@program.offset_of(type.sizeof_type, ivar_index)
# and
program.instance_offset_of(type.sizeof_type, ivar_index)
```

we can assign that to a variable and do all of the `expanded = ...; expanded.type = ...; node.expanded = ...` outside of the `if`s.